src/azure-cli-core/azure/cli/core/commands/arm.py

```diff
@@ -89,25 +89,58 @@ def deployment_validate_table_format(result):
+    if has_grandchild:
+        parent = '{type}/{name}/providers/{child_namespace}/{child_type}/{child_name}/providers/{grandchild_namespace}'.format(**kwargs)  # pylint: disable=line-too-long
+    elif has_child:
+        parent = '{type}/{name}/providers/{child_namespace}'.format(**kwargs)
+    parent = parent.replace('providers/None', '')
+    parent = '{}/'.format(parent) if parent and not parent.endswith('/') else parent
+
+    kwargs['resource_parent'] = parent
+    kwargs['resource_namespace'] = resource_namespace
+    kwargs['resource_type'] = resource_type
+    kwargs['resource_name'] = resource_name
+    return kwargs
+
+
 def resource_id(**kwargs):
     '''Create a valid resource id string from the given parts
     The method accepts the following keyword arguments:
-        - subscription      Subscription id
-        - resource_group    Name of resource group
-        - namespace         Namespace for the resource provider (i.e. Microsoft.Compute)
-        - type              Type of the resource (i.e. virtualMachines)
-        - name              Name of the resource (or parent if child_name is also specified)
-        - child_type        Type of the child resource
-        - child_name        Name of the child resource
-        - grandchild_type   Type of the grandchild resource
-        - grandchild_name   Name of the grandchild resource
+        - subscription          Subscription id
+        - resource_group        Name of resource group
+        - namespace             Namespace for the resource provider (i.e. Microsoft.Compute)
+        - type                  Type of the resource (i.e. virtualMachines)
+        - name                  Name of the resource (or parent if child_name is also specified)
+        - child_namespace       Namespace for the child resoure (optional)
+        - child_type            Type of the child resource
+        - child_name            Name of the child resource
+        - grandchild_namespace  Namespace for the grandchild resource (optional)
+        - grandchild_type       Type of the grandchild resource
+        - grandchild_name       Name of the grandchild resource
     '''
     rid = '/subscriptions/{subscription}'.format(**kwargs)
     try:
         rid = '/'.join((rid, 'resourceGroups/{resource_group}'.format(**kwargs)))
         rid = '/'.join((rid, 'providers/{namespace}'.format(**kwargs)))
         rid = '/'.join((rid, '{type}/{name}'.format(**kwargs)))
+        try:
+            rid = '/'.join((rid, 'providers/{child_namespace}'.format(**kwargs)))
+        except KeyError:
+            pass
         rid = '/'.join((rid, '{child_type}/{child_name}'.format(**kwargs)))
+        try:
+            rid = '/'.join((rid, 'providers/{grandchild_namespace}'.format(**kwargs)))
+        except KeyError:
+            pass
         rid = '/'.join((rid, '{grandchild_type}/{grandchild_name}'.format(**kwargs)))
     except KeyError:
```

Comment:
I use the outer try/except because that's what was originally in place.  It "just works" to stop constructing the ID string at the appropriate length. I had to add the inner try/excepts (which are super ugly imo) because the provider namespace may or may not exist.